### PR TITLE
Test foolscap to http upgrade

### PR DESCRIPTION
**Pull request by itamarst**
Fixes https://tahoe-lafs.org/trac/tahoe-lafs/ticket/4047

My initial thought was to have an integration test too, but from the outside it's impossible to tell whether HTTP or Foolscap was used and we don't have a disable-Foolscap option, so I didn't in the end.

### CODE REVIEW

File: src/allmydata/test/test_storage_client.py [hunk 1]
```diff
@@ -645,6 +646,43 @@ def test_static_servers(self):
...
+        """
+        When an announcement is initially Foolscap but then switches to HTTP,
+        HTTP is used, assuming HTTP is enabled.
+        """
+        tub_maker = lambda _: new_tub()
+        config = config_from_string(
+            "/dev/null", "", "[client]\nforce_foolscap = False\n"
+        )
+        broker = StorageFarmBroker(True, tub_maker, config)
+        broker.startService()
```
Comment (exarkun) [inline src/allmydata/test/test_storage_client.py:659]:
```suggestion
        broker.startService()
        self.addCleanup(broker.stopService)
```

File: src/allmydata/test/test_storage_client.py [hunk 2]
```diff
@@ -645,6 +646,43 @@ def test_static_servers(self):
...
+        http_announcement[ANONYMOUS_STORAGE_NURLS] = {f"pb://{ones}@nowhere/fake2#v=1"}
+        broker._got_announcement(key_s, http_announcement)
+        self.assertFalse(initial_service.running)
+        self.assertEqual(initial_service.parent, None)
+        new_service = broker.servers[key_s]
+        self.assertIsInstance(new_service, HTTPNativeStorageServer)
+        self.assertTrue(new_service.running)
+        self.assertIdentical(new_service.parent, broker)
+
+        return broker.stopService()
```
Comment (exarkun) [inline src/allmydata/test/test_storage_client.py:684]:
```suggestion
```

**Review by exarkun — APPROVED** [review]
Nice, thanks.  Just a pair of inline comments.  Address as you like and then merge.

**Comment by itamarst** [comment]
I don't see how the integration test failure can be blamed on this PR, given it only changes unit tests. PyPy failures might be related though, so will look into that.

**Comment by itamarst** [comment]
Oh, but the failure (reproducible) is in a test module that wasn't touched by this PR. So probably due to new release of Hypothesis being stricter or something. Will validate and open ticket if so.

**Comment by itamarst** [comment]
All the failures are unrelated to this PR, either due to https://tahoe-lafs.org/trac/tahoe-lafs/ticket/4063 or clearly unrelated integration test failure (this only modifies a test!), so going to merge.

**Pull request merged**